### PR TITLE
accelerator: build components as dso's by default

### PR DESCRIPTION
**Pull request by hppritcha**
related to #12036

### CODE REVIEW

**Comment by hppritcha** [comment]
bot:ompi:retest

strange failure of make check shouldn't be impacted by this PR.  rerunning

**Comment by hppritcha** [comment]
bot:aws:retest

**Comment by jsquyres** [comment]
bot:aws:retest

File: config/opal_mca.m4 [hunk 1]
```diff
@@ -176,14 +179,16 @@ of type-component pairs.  For example, --enable-mca-no-build=pml-ob1])
...
+        DSO_accelerator=1
+        msg="accelerator framework only"
     elif test "$enable_mca_dso" = "no"; then
         DSO_all=0
         msg=none
     elif test "$enable_mca_dso" = "yes"; then
         DSO_all=1
         msg=all
     else
+        enable_mca_dso=$enable_mca_dso",accelerator"
```
Comment (jsquyres) [inline config/opal_mca.m4:191]: I don't think we want to always automatically add `accelerator` to whatever list the user supplies. I think that if the user supplies a list, we should do exactly that list.

File: config/opal_mca.m4 [hunk 2]
```diff
@@ -176,14 +179,16 @@ of type-component pairs.  For example, --enable-mca-no-build=pml-ob1])
         msg="none (dlopen disabled)"
     elif test -z "$enable_mca_dso"; then
         DSO_all=0
-        msg=default
+        DSO_accelerator=1
```
Comment (jsquyres) [inline config/opal_mca.m4:182]: Should this be:
```suggestion
enable_mca_dso=accelerator
```
?

File: config/opal_mca.m4 [hunk 3]
```diff
@@ -176,14 +179,16 @@ of type-component pairs.  For example, --enable-mca-no-build=pml-ob1])
...
+        msg="accelerator framework only"
```
Comment (jsquyres) [inline config/opal_mca.m4:183]: Maybe do this:
```suggestion
        msg="accelerator framework only (default)"
```

File: config/opal_mca.m4 (hunk 1 above)
Comment (hppritcha) [reply config/opal_mca.m4:191]: > I don't think we want to always automatically add `accelerator` to whatever list the user supplies. I think that if the user supplies a list, we should do exactly that list.

okay makes sense

File: config/opal_mca.m4 (hunk 2 above)
Comment (hppritcha) [reply config/opal_mca.m4:182]: that makes things a little complicated since we'd need to repeat the logic at 
```
for item in $enable_mca_dso; do
            AS_VAR_SET([AS_TR_SH([DSO_$item])], [1])
            msg="$item $msg"
        done
        ```
a second time or move that out of the if block and put in a separate one below.

File: config/opal_mca.m4 (hunk 3 above)
Comment (edgargabriel) [reply config/opal_mca.m4]: don't we have to add btl/smcuda and rcache/gpusm,, rcache/rgpusm also to the list of components that are compiled as dso's by default? Otherwise they will pull in the cuda libraries as well

Comment (hppritcha) [reply config/opal_mca.m4]: > don't we have to add btl/smcuda and rcache/gpusm,, rcache/rgpusm also to the list of components that are compiled as dso's by default? Otherwise they will pull in the cuda libraries as well

I'm okay with this although I thought we only wanted this DSO by default for accelerator framework.

Comment (edgargabriel) [reply config/opal_mca.m4]: Lets discuss this in the meeting tomorrow. I think it probably comes down the precise goal that we want to achieve with making the accelerator components built as dsos

Comment (hppritcha) [reply config/opal_mca.m4]: hmm, i think something in the configury is broken.  I seem to only be able to set DSO builds at the framework level and not at the individual component level.  there are a number of differences between opal_mca.m4 in the main/5.0.0 and 4.1.x.  investigating...

**Comment by hppritcha** [comment]
okay, i've added smcuda, accelerator (so all components), and rcache-gpusm rcache-rgpusm but still see libcuda linked in to opal and ompi libs and hence to commands like ompi_info.  digging...

**Comment by hppritcha** [comment]
oh i had a typo in the config file.  ignore last comment.

**Comment by hppritcha** [comment]
i still have no idea why the make check ```opal_condition``` would fail with this PR.

**Comment by hppritcha** [comment]
@jsquyres  @edgargabriel  please recheck

**Review by edgargabriel — APPROVED** [review]
I can confirm that these components are now built as dso's by default.

You need to rebase your branch however, it is missing the Makefile.am fixes to btl/smcuda and the two rcache components, and hence doesn't compile.

File: config/opal_mca.m4 [hunk 4]
```diff
@@ -167,29 +167,34 @@ of type-component pairs.  For example, --enable-mca-no-build=pml-ob1])
...
     elif test "$enable_mca_dso" = "no"; then
         DSO_all=0
         msg=none
     elif test "$enable_mca_dso" = "yes"; then
         DSO_all=1
         msg=all
     else
+        if test -z "$enable_mca_dso"; then
+            enable_mca_dso="accelerator,btl-smcuda,rcache-gpusm,rcache-rgpusm"
+            AC_MSG_NOTICE([building accelerator framework compnents as DSOs (default)])
```
Comment (jsquyres) [inline config/opal_mca.m4:189]: Putting this `AC_MSG_NOTICE` here will interrupt the `AC_MSG_CHECKING` / `AC_MSG_RESULT`.

Maybe you should just assign `msg` here with some kind of suffix to put after the list of components in the `AC_MSG_RESULT`...?  Perhaps something like (just typed here in github; not tested):

```sh
        msg=
        if test -z "$enable_mca_dso"; then
            # We set all accelerator-related components to build as DSO by default.
            # See ...URL of github issue or something with more explanation...
            enable_mca_dso="accelerator,btl-smcuda,rcache-gpusm,rcache-rgpusm"
            msg="(default)"
        fi
        DSO_all=0
        ifs_save="$IFS"
        IFS="${IFS}$PATH_SEPARATOR,"
        for item in $enable_mca_dso; do
            AS_VAR_SET([AS_TR_SH([DSO_$item])], [1])
            str="`echo DSO_$item=1 | sed s/-/_/g`"
            eval $str
            msg="$item $msg"
        done
        IFS="$ifs_save"
    fi
    AC_MSG_RESULT([$msg])

```

Comment (hppritcha) [reply config/opal_mca.m4:189]: that sounds like a good idea.  fixing.

File: config/opal_mca.m4 [hunk 5]
```diff
@@ -167,29 +167,34 @@ of type-component pairs.  For example, --enable-mca-no-build=pml-ob1])
...
+        fi
         DSO_all=0
         ifs_save="$IFS"
         IFS="${IFS}$PATH_SEPARATOR,"
         msg=
         for item in $enable_mca_dso; do
-            AS_VAR_SET([AS_TR_SH([DSO_$item])], [1])
+            str="`echo DSO_$item=1 | sed s/-/_/g`"
+            eval $str
```
Comment (jsquyres) [inline config/opal_mca.m4:197]: Why did you remove the `AS_VAR_SET`?  Was this code broken for `foo-var` kinds of `$item` values?

Comment (hppritcha) [reply config/opal_mca.m4:197]: yes it was

Comment (hppritcha) [reply config/opal_mca.m4:197]: one of the reasons for dalcinl frustrations with building certain components (as opposed to frameworks for which the previous logic worked) as dso.s. i forgot which pr/issue that was though where he expressed that frustration with main and 5.0.x

**Comment by hppritcha** [comment]
i'm going to push a temporary commit to undo the default dso module thing to see if that's somehow causing the make check to fail.

**Comment by hppritcha** [comment]
okay the problem is that opal_init call in the opal_conditional test fails in the make distcheck build.  can't find an accelerator component.

**Comment by hppritcha** [comment]
This is rather annoying.  I've narrowed the problem down to the fact that when we try to select an accelerator component, if none are found, ```opal_init``` returns a fatal error.  Now, when ```make check``` is run but ```make install``` has not been previously run, then the framework can't find a valid accelerator component and thus returns error.  so any make check program that calls ```opal_init``` will error out unless ```make install``` had been previously run.  doesn't have to do with the dist tarball itself.

@jsquyres 
@edgargabriel 

thoughts?

**Comment by hppritcha** [comment]
i forgot that we can specify specific accelerator components to build as DSO's so trying that route.

**Comment by edgargabriel** [comment]
In theory there is no reason to have the null component as a dso, that could be part statically compiled, and that might resolve your problem

**Comment by hppritcha** [comment]
that's what I'm trying.

note this behavior in the accel base select component may need a better fix.  For example, if someone configures with ```--enable-mca-dso``` then does make check, they will see the same failures as this PR was exhibiting.  kind of an edge case but still...

**Comment by hppritcha** [comment]
@jsquyres please review again

**Comment by hppritcha** [comment]
@jsquyres ping, i'd like to pr over to 5.0.x for the 5.0.1 release

File: config/opal_mca.m4 (hunk 5 above)
Comment (jsquyres) [reply config/opal_mca.m4:198]: Can you provide an example of what this `AS_VAR_SET` + `AS_TR_SH` was handling wrong?  The point of these macros is to ensure safe handling of indirection and not-valid-for-variable-name characters.  E.g., `AS_TR_SH` handles the transmorgification of `-` to `_`.

Comment (hppritcha) [reply config/opal_mca.m4:198]: https://unix.stackexchange.com/questions/428880/list-of-acceptable-initial-characters-for-a-bash-variable

Comment (hppritcha) [reply config/opal_mca.m4:198]: you are wrong.  - remains - when run through AS_TR_SH

Comment (hppritcha) [reply config/opal_mca.m4:198]: well i retract that statement.  doing more testing.

File: config/opal_mca.m4 [hunk 6]
```diff
@@ -167,29 +167,35 @@ of type-component pairs.  For example, --enable-mca-no-build=pml-ob1])
...
         DSO_all=0
         msg=none
     elif test "$enable_mca_dso" = "yes"; then
         DSO_all=1
         msg=all
     else
+        msg=
+       if test -z "$enable_mca_dso"; then
+           enable_mca_dso="accelerator-cuda,accelerator-rocm,accelerator-ze,btl-smcuda,rcache-gpusm,rcache-rgpusm"
+           msg="(default)"
```
[the remark this answers is not in the document]
Comment (hppritcha) [reply config/opal_mca.m4]: actually i don't like this because it will display them twice so leaving with (default)

**Comment by hppritcha** [comment]
@open-mpi/cuda could someone see what's going on with ompi_NVIDIA CI?

**Comment by hppritcha** [comment]
@jsquyres ping

File: config/opal_mca.m4 (hunk 6 above)
[the remark this answers is not in the document]
Comment (jsquyres) [reply config/opal_mca.m4:190]: Ah, I see -- we prepend $msg with each of the component/framework names.  Got it.  👍

**Review by jsquyres — APPROVED** [review]
